src/ImageProcessor/Imaging/ResizeHelper.cs

```diff
@@ -0,0 +1,394 @@
+
+            // Replace the size to match the rectangle.
+            return (new Size(targetWidth, targetHeight), new Rectangle(0, 0, targetWidth, targetHeight));
+        }
+
+        private static (Size, Rectangle) CalculatePadRectangle(
+            Size sourceSize,
+            ResizeLayer options,
+            int width,
+            int height)
+        {
+            float ratio;
+            int sourceWidth = sourceSize.Width;
+            int sourceHeight = sourceSize.Height;
+
+            int targetX = 0;
+            int targetY = 0;
+            int targetWidth = width;
+            int targetHeight = height;
+
+            // Fractional variants for preserving aspect ratio.
+            float percentHeight = Math.Abs(height / (float)sourceHeight);
+            float percentWidth = Math.Abs(width / (float)sourceWidth);
+
+            if (percentHeight < percentWidth)
+            {
+                ratio = percentHeight;
+                targetWidth = (int)Math.Round(sourceWidth * percentHeight);
+
+                switch (options.AnchorPosition)
+                {
+                    case AnchorPosition.Left:
+                    case AnchorPosition.TopLeft:
+                    case AnchorPosition.BottomLeft:
+                        targetX = 0;
+                        break;
+                    case AnchorPosition.Right:
+                    case AnchorPosition.TopRight:
+                    case AnchorPosition.BottomRight:
+                        targetX = (int)Math.Round(width - (sourceWidth * ratio));
+                        break;
+                    default:
+                        targetX = (int)Math.Round((width - (sourceWidth * ratio)) / 2F);
+                        break;
+                }
+            }
+            else
+            {
+                ratio = percentWidth;
+                targetHeight = (int)Math.Round(sourceHeight * percentWidth);
+
+                switch (options.AnchorPosition)
+                {
+                    case AnchorPosition.Top:
+                    case AnchorPosition.TopLeft:
+                    case AnchorPosition.TopRight:
+                        targetY = 0;
+                        break;
+                    case AnchorPosition.Bottom:
+                    case AnchorPosition.BottomLeft:
+                    case AnchorPosition.BottomRight:
+                        targetY = (int)Math.Round(height - (sourceHeight * ratio));
+                        break;
+                    default:
+                        targetY = (int)Math.Round((height - (sourceHeight * ratio)) / 2F);
+                        break;
+                }
+            }
+
+            // Target image width and height can be different to the rectangle width and height.
+            return (new Size(width, height), new Rectangle(targetX, targetY, targetWidth, targetHeight));
+        }
+
+        private static void ThrowInvalid(string message) => throw new InvalidOperationException(message);
```

Comment:
This seems like a silly helper method that will actually screw up the stack trace and line numbers.

Reply:
Actually static helper voids are very useful for code generation, especially on hotpaths where they allow inlining of calling methods. (this isn't one but every little helps.). By making them static you remove runtime overhead.

Reply:
But that's not the case here and it still messes with the stack trace: https://stackoverflow.com/a/55486648/96175.

Reply:
That's not the same thing. You're comparing a throw helper in a different class to a private static member of the same class. 

The example given would require the `DebuggerStepThroughAttribute` to be present in order not to disrupt the stack trace.

Reply:
Just to clarify, I meant the call stack on the exception that's thrown... That would still include the `ThrowInvalid` method/line number, even when it's marked with the `DebuggerStepThroughAttribute` (or even `DebuggerHiddenAttribute`).

So, if removing the exception handling from the `CalculateTargetLocationAndBounds` method would indeed cause it to be inlined and enhance performance for this hotpath, that would certainly be a good justification 😉